typed-racket-lib/typed-racket/private/type-contract.rkt

```diff
@@ -954,17 +950,98 @@
+    ;;                                  (-> (-> Type Contract)
+    ;;                                      (-> (Listof Type) Contract))))
+    ;; a TC-Quotient (type-contract quotient) is a function that represents
+    ;;  a "set of types" that can be enforced with one contract.
+    ;; If `tc` is a TC-Quotient:
+    ;; - `(tc t0 t1)` returns #true iff types `t0` and `t1` are in the same set
+    ;; - `((tc t->sc) ts)` returns the canonical contract
+    ;;   for `ts`, assuming the `ts` are all in the same set
+    ;;
+    ;; make-simple-tc-quotient : (-> (-> Type Type (U #f (-> Type Contract)) (U Boolean Contract)) TC-Quotient)
+    ;; Make a TC-Quotient from a function that returns a canonical contract
+    ;;  for two types
+    (define (make-simple-tc-quotient q2)
+      (case-lambda
+       [(t0 t1)
+        (q2 t0 t1 #f)]
+       [(t->sc)
+        (λ (ts)
+          (cond
+           [(null? ts)
+            (raise-argument-error 'simple-tc-quotient "non-empty list" 0 ts t->sc)]
+           [(null? (cdr ts))
+            ;; Don't generalize a single type
+            (t->sc (car ts))]
+           [else
+            ;; Assumes: for all `t0` in `ts`,
+            ;;          for all `t1` in `ts`,
+            ;;          `(q2 t0 t1 t->sc)` returns the same contract
+            (q2 (first ts) (second ts) t->sc)]))]))
+    (define TC-QUOTIENTS
+      (map make-simple-tc-quotient (list
+        (λ (t0 t1 maybe-t->sc)
+          (match* (t0 t1)
+           [((or (Immutable-HashTable: k v)
+                 (Mutable-HashTable: k v)
+                 (Weak-HashTable: k v))
+             (or (Immutable-HashTable: k v)
+                 (Mutable-HashTable: k v)
+                 (Weak-HashTable: k v)))
+            (if maybe-t->sc
+              (hash/sc (maybe-t->sc k) (maybe-t->sc v))
+              #true)]
+           [(_ _) #f]))
+        (λ (t0 t1 maybe-t->sc)
+          (match* (t0 t1)
+           [((HeterogeneousVector: ts)
+             (HeterogeneousVector: ts))
+            (if maybe-t->sc
+              (apply vector/sc (map maybe-t->sc ts))
+              #true)]
+           [(_ _) #f]))
+        (λ (t0 t1 maybe-t->sc)
+          (match* (t0 t1)
+           [((Vector: t)
+             (Vector: t))
+            (if maybe-t->sc
+              (vectorof/sc (maybe-t->sc t))
+              #true)]
+           [(_ _) #f])))))
+    ;; types->contracts : TC-Quotient (Listof Type) (-> Type Contract) -> (Listof Contract)
+    ;; Groups a list of types into equivalence classes using `q`,
+    ;;  return a list containing the canonical elements of each equivalence class
+    ;; The output list may have fewer elements than the input.
+    (define (types->contracts q ts t->sc)
+      (match ts
+       ['()
+        '()]
+       [(cons t ts)
+        (define (=t t+)
+          (q t t+))
+        (define-values [qts other] (partition =t ts))
+        (cons ((q t->sc) (cons t qts)) (types->contracts q other t->sc))]))
+    (λ (ts t->sc)
```

Comment:
Oops, having just one `t->sc`continuation is a design bug.
If you wanted to merge two arrow contracts, `(-> A B)` and `(-> C D)`, you'd want something like `(-> (t->sc/neg (merge A C)) (t->sc (merge B D)))`.

I'll change / specialize `type-contract-generalize` later tonight.